utils/env_preferences/src/linux.rs

```diff
@@ -0,0 +1,112 @@
+    pub enum LocaleCategory {
+        Character,
+        Number,
+        Time,
+        Collate,
+        Monetary,
+        Messages,
+        Paper,
+        Name,
+        Address,
+        Telephone,
+        Measurement,
+        Identification,
+        All,
+        Other(String),
+    }
+
+    impl FromStr for LocaleCategory {
+        type Err = ();
+
+        fn from_str(s: &str) -> Result<Self, Self::Err> {
+            match s {
+                "LC_CTYPE" => Ok(LocaleCategory::Character),
+                "LC_NUMERIC" => Ok(LocaleCategory::Number),
+                "LC_TIME" => Ok(LocaleCategory::Time),
+                "LC_COLLATE" => Ok(LocaleCategory::Collate),
+                "LC_MONETARY" => Ok(LocaleCategory::Monetary),
+                "LC_MESSAGES" => Ok(LocaleCategory::Messages),
+                "LC_PAPER" => Ok(LocaleCategory::Paper),
+                "LC_NAME" => Ok(LocaleCategory::Name),
+                "LC_ADDRESS" => Ok(LocaleCategory::Address),
+                "LC_TELEPHONE" => Ok(LocaleCategory::Telephone),
+                "LC_MEASUREMENT" => Ok(LocaleCategory::Measurement),
+                "LC_IDENTIFICATION" => Ok(LocaleCategory::Identification),
+                "LC_ALL" => Ok(LocaleCategory::All),
+                _ => Ok(LocaleCategory::Other(s.to_string())),
+            }
+        }
+    }
+
+    pub fn get_locales() -> HashMap<LocaleCategory, String> {
+        let mut locale_map = HashMap::new();
+
+        // SAFETY: Safety is ensured because we pass a `NULL` pointer and retrieve the locale there is
+        // no subsequent calls for `setlocale` which could change the locale of this particular thread
+        let locales_ptr = unsafe { setlocale(LC_ALL, ptr::null()) };
+
+        if locales_ptr.is_null() {
+            locale_map.insert(LocaleCategory::All, "C".to_string());
+            return locale_map;
+        }
+
+        // SAFETY: A valid `NULL` terminator is present which is a requirement of `from_ptr`
+        let locales_cstr = unsafe { CStr::from_ptr(locales_ptr) };
+
+        if let Ok(locales_str) = locales_cstr.to_str() {
+            let locale_pairs = locales_str.split(';');
+
+            // To handle cases in case a single locale is returned or a list of locale
+            if locale_pairs.clone().count() == 1 {
+                locale_map.insert(LocaleCategory::All, "C".to_string());
+            } else {
+                for locale_pair in locale_pairs {
+                    let mut parts = locale_pair.split('=');
+                    if let (Some(key), Some(value)) = (parts.next(), parts.next()) {
+                        if let Ok(category) = LocaleCategory::from_str(key) {
+                            locale_map.insert(category, value.to_string());
+                        }
+                    }
+                }
+            }
+        }
+
+        locale_map
+    }
+
+    pub fn get_system_calendars() -> impl Iterator<Item = (Cow<'static, str>, Cow<'static, str>)> {
```

Comment:
why Cow? And Why Iterator? I'd return Vec<> of Strings. It looks like premature optimization.